utils/tfw_logger.cc

```diff
@@ -0,0 +1,346 @@
+		INT_CASE(TFW_MMAP_LOG_VERSION, clickhouse::ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_STATUS, clickhouse::ColumnUInt16, uint16_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_CONT_LEN, clickhouse::ColumnUInt32, uint32_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_TIME, clickhouse::ColumnUInt32, uint32_t);
+		INT_CASE(TFW_MMAP_LOG_DROPPED, clickhouse::ColumnUInt64, uint64_t);
+
+		case TFW_MMAP_LOG_VHOST:
+		case TFW_MMAP_LOG_URI:
+		case TFW_MMAP_LOG_REFERER:
+		case TFW_MMAP_LOG_USER_AGENT:
+			if (!TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+				(*block)[ind]->As<clickhouse::ColumnString>()->Append(
+					std::move(std::string("")));
+				break;
+			}
+			len = *((uint16_t *)p);
+			if (len + 2 > size)
+				return -1;
+			(*block)[ind]->As<clickhouse::ColumnString>()->Append(
+				std::move(std::string(p + 2, len)));
+			len += 2;
+			break;
+		default:
+			throw Except("Unknown field type: {}", i);
+		}
+
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+			p += len;
+			size -= len;
+		}
+	}
+
+	return p - data;
+}
+
+void
+callback(const char *data, int size, void *private_data)
+{
+	TfwClickhouse *clickhouse = (TfwClickhouse *)private_data;
+	TfwBinLogEvent *event;
+	const char *p = data;
+	int r;
+
+	dbg_hexdump(data, size);
+
+	do {
+		if (size < (int)sizeof(TfwBinLogEvent))
+			return;
+
+		event = (TfwBinLogEvent *)p;
+
+		switch (event->type) {
+		case TFW_MMAP_LOG_TYPE_ACCESS:
+			r = read_access_log_event(p, size, clickhouse);
+			if (r < 0)
+				return;
+			size -= r;
+			p += r;
+			break;
+		default:
+			throw Except("Unsupported log type: {}",
+				     (unsigned int)event->type);
+			return;
+		}
+	} while (size > 0);
+
+	clickhouse->commit();
+}
+
+void
+run_thread(int ncpu, int fd, std::string host, std::promise<void> promise)
+try {
+	cpu_set_t cpuset;
+	pthread_t current_thread = pthread_self();
+
+	TfwClickhouse clickhouse(host, TABLE_NAME, make_block);
+
+	TfwMmapBufferReader mbr(ncpu, fd, &clickhouse, callback);
+
+	CPU_ZERO(&cpuset);
+	CPU_SET(mbr.get_cpu_id(), &cpuset);
+
+	assert(pthread_setaffinity_np(current_thread,
+				      sizeof(cpu_set_t), &cpuset) == 0);
+
+	mbr.run(&stop_flag);
+
+	promise.set_value();
+}
+catch (...) {
+	promise.set_exception(std::current_exception());
+}
+
+void
+sig_handler([[maybe_unused]] int  sig_num) noexcept
+{
+	stop_flag = true;
+}
+
+void
+set_sig_handlers()
+{
+	struct sigaction sa;
+
+	sigemptyset(&sa.sa_mask);
+	sigaddset(&sa.sa_mask, SIGHUP);
+	sigaddset(&sa.sa_mask, SIGINT);
+	sigaddset(&sa.sa_mask, SIGTERM);
+
+	sa.sa_handler = sig_handler;
+	sa.sa_flags = SA_RESTART;
+
+	sigaction(SIGHUP, &sa, NULL);
+	sigaction(SIGINT, &sa, NULL);
+	sigaction(SIGTERM, &sa, NULL);
+}
+
+int
+main(int argc, char* argv[])
+try {
+	std::vector<std::thread> thrs;
+	std::vector<std::future<void>> futures;
+	unsigned int i;
+	long int cpu_cnt;
+	int fd;
+
+	po::options_description desc{"Usage: tfw_logger [options] <host>"};
+	desc.add_options()
```

Comment:
I catch `Unhandled error: DB::Exception: default: Authentication failed: password is incorrect, or there is no user with such name.` It seems that we should have option password

Reply:
@krizhanovsky do we need password support?

Reply:
Added user/password support.